pkg/serviceaccounttoken/secret.go

```diff
@@ -27,12 +28,27 @@ const (
 // as long as it can wrap it in a simplified lambda with this signature.
 type secretLister func(namespace string, selector labels.Selector) ([]*v1.Secret, error)
 
+var lockMap sync.Map
+
+func getLock(key string) *sync.Mutex {
+	actual, _ := lockMap.LoadOrStore(key, &sync.Mutex{})
+	return actual.(*sync.Mutex)
+}
+
 // EnsureSecretForServiceAccount gets or creates a service account token Secret for the provided Service Account.
 // For k8s <1.24, the secret is automatically generated for the service account. For >=1.24, we need to generate it explicitly.
 func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrollers.SecretCache, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
 	if sa == nil {
 		return nil, fmt.Errorf("could not ensure secret for invalid service account")
 	}
+	logrus.Tracef("EnsureSecretForServiceAccount for %s/%s", sa.GetNamespace(), sa.GetName())
+
+	// Lock avoids multiple calls to this func at the same time and creation of same resources multiple times
+	// Mutex is a addition to the Lease, it helps sync within the pod and avoid multiple Lease waits from the same pod
+	mutex := getLock(fmt.Sprintf("%v-%v", sa.Namespace, sa.Name))
```

Comment:
I might be off-base here, but should we also include the "cluster id" as part of the lock? I'm thinking that in theory if this is called for the same user, this will lock for all clusters (since it's the same name and namespace), but in theory this wait wouldn't be useful since the create will only happen for one cluster.

Reply:
I'm not sure how we'd determine the cluster here?

Reply:
The ID would have to be passed to `EnsureSecretForServiceAccount` as a new parameter I guess, as the cluster is "encoded" inside of `secretsCache` and `clientSet` but it is not exposed (and those objects might not even be aware of the Rancher ID).

Reply:
We could do a new parameter or allow the context to contain an optional prefix that could be used as a prefix to the lock key. Either way, unless I'm missing something we do need something here or this could lock up concurrent requests to different clusters.